prerender-until-script.md

```diff
@@ -0,0 +1,123 @@
+
+## Key Scenarios / Use Cases
+
+*   **Content sites with heavy interactive elements:** A news article or blog post could be prerendered to display the text content instantly, while the JavaScript for comments sections, social media widgets, and other interactive elements is deferred until the user activates the page.
+*   **E-commerce product pages:** A product page could be prerendered to show the main product image and description, while scripts for engagement calculation and conversion tracking are paused until the user navigates to the page.
+
+## Proposed Solution
+
+We propose a new action, `prerender_until_script`, which can be used within a `<script type="speculationrules">` element. When the browser processes a speculation rule with this action, it will initiate a prerender that:
+
+1.  Fetches the document.
+2.  Parses the HTML and constructs the DOM. The user agent's preload scanner can discover and begin downloading subresources (e.g., stylesheets, images) referenced in the document.
+3.  Renders the page. If a parser-blocking `<script>` element is encountered, rendering will be paused at that point until activation.
+4.  Pauses all script execution. This includes `<script>` blocks, timers, and promise resolutions.
+5.  When the user navigates to the prerendered page, the page is activated, and all paused scripts are executed in order.
+
+### Example
+
+```html
+<!-- initiator.html -->
+<script type="speculationrules">
+{
+  "prerender_until_script": [
+    {
+      "source": "list",
+      "urls": ["/heavy-script-page.html"]
+    }
+  ]
+}
+</script>
+
+<a href="/heavy-script-page.html">Link to a page with heavy scripts</a>
+```
+
+```html
+<!-- heavy-script-page.html -->
+<script defer scr="deferred-script.js" id="deferred-script"></script>
+<script async src="async-script.js" id="async-script"></script>
+<img src="image1.jpg" id="first-image"/>
+<script id="first-blocking">
+{
+  // Heavy script... 
+}
+</script>
+<img src="image2.jpg"/>
+
+```
+
+In this example, `/heavy-script-page.html` will be prerendered, but its scripts will not run until the user clicks the link and the page is activated.
+
+More specifically, the user agent processes the page as follows:
+
+1.  The `deferred-script` script is parsed and added to the queue of scripts that will be executed after page parsing is complete. Parsing then resumes.
+2.  The `async-script` script is parsed and added to the queue of scripts that will be executed after prerender activation. Parsing then resumes.
+3.  The `first-image` element is parsed, `image1.jpg` is fetched and loaded, and parsing resumes.
+4.  When the `first-blocking` script is parsed, the parser pauses and waits for prerender activation.
+5.  While waiting for activation, the user agent may optionally prefetch `image2.jpg`.
+6.  The prerendered document is activated.
+7.  The parsing of `first-blocking` script resumes, and the script is executed.
```

Comment:
Now continue the story and tell me about when and in what order `deferred-script` and `async-script` are executed!

Reply:
Added! :D